File: src/strategies/early-bird.js
// early-bird-5m — signal logic for 5-min binary markets
// Indicators: Wilder RSI(14) on 1m candles + ATR(14) + price divergence

export function computeRSI(closes, period = 14) {
  if (closes.length < period + 1) return null;

  const changes = [];
  for (let i = 1; i < closes.length; i++) {
    changes.push(closes[i] - closes[i - 1]);
  }

  let avgGain = 0, avgLoss = 0;
  for (let i = 0; i < period; i++) {
    if (changes[i] > 0) avgGain += changes[i];
    else avgLoss -= changes[i];
  }
  avgGain /= period;
  avgLoss /= period;

  for (let i = period; i < changes.length; i++) {
    const gain = changes[i] > 0 ? changes[i] : 0;
    const loss = changes[i] < 0 ? -changes[i] : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
  }

  if (avgLoss === 0) return 100;
  return 100 - 100 / (1 + avgGain / avgLoss);
}

export function computeATR(candles, period = 14) {
  if (candles.length < 2) return null;

  const trs = [];
  for (let i = 1; i < candles.length; i++) {
    const { high, low } = candles[i];
    const prevClose = candles[i - 1].close;
    trs.push(Math.max(high - low, Math.abs(high - prevClose), Math.abs(low - prevClose)));
  }

  const slice = trs.slice(-period);
  return slice.reduce((a, b) => a + b, 0) / slice.length;
}

// Returns { outcome: 'UP' | 'DOWN' } or null
// - spotPrice: current Binance spot price
// - priceToBeat: the market's resolution threshold
// - rsi: Wilder RSI value
// - atr: Average True Range
// - upPrice / downPrice: current CLOB prices for each outcome (0-1)
export function generateSignal({ spotPrice, priceToBeat, rsi, atr, upPrice, downPrice }) {
  if (rsi === null) return null;

  // Skip if market is flat (< 0.005% ATR) or too chaotic (> 0.5% ATR)
  if (atr !== null) {
    const atrPct = (atr / spotPrice) * 100;
    if (atrPct < 0.005 || atrPct > 0.5) return null;
  }

  // "Up or Down" markets: no price threshold, use pure RSI + outcome pricing
  if (priceToBeat === null) {
    if (rsi >= 58 && upPrice !== null && upPrice < 0.52) return { outcome: 'UP' };
    if (rsi <= 42 && downPrice !== null && downPrice < 0.52) return { outcome: 'DOWN' };
    return null;
  }

  // Markets with explicit threshold (e.g. "Will BTC be above $82,000?")
  const divergence = (spotPrice - priceToBeat) / priceToBeat;

  // BUY UP: momentum bullish + spot trending above threshold + UP side is cheap
  if (rsi >= 58 && divergence > 0.0005 && upPrice !== null && upPrice < 0.52) {
    return { outcome: 'UP' };
  }

  // BUY DOWN: momentum bearish + spot trending below threshold + DOWN side is cheap
  if (rsi <= 42 && divergence < -0.0005 && downPrice !== null && downPrice < 0.52) {
    return { outcome: 'DOWN' };
  }

  return null;
}

File: backend/strategies/btc5m/early_bird.py
# early-bird-5m — signal logic for 5-min binary markets
# Indicators: Wilder RSI(14) on 1m candles + ATR(14) + price divergence


def compute_rsi(closes: list[float], period: int = 14) -> float | None:
    if len(closes) < period + 1:
        return None

    changes = [closes[i] - closes[i - 1] for i in range(1, len(closes))]

    avg_gain = avg_loss = 0.0
    for i in range(period):
        if changes[i] > 0:
            avg_gain += changes[i]
        else:
            avg_loss -= changes[i]
    avg_gain /= period
    avg_loss /= period

    for i in range(period, len(changes)):
        gain = changes[i] if changes[i] > 0 else 0.0
        loss = -changes[i] if changes[i] < 0 else 0.0
        avg_gain = (avg_gain * (period - 1) + gain) / period
        avg_loss = (avg_loss * (period - 1) + loss) / period

    if avg_loss == 0:
        return 100.0
    return 100 - 100 / (1 + avg_gain / avg_loss)


def compute_atr(candles: list[dict], period: int = 14) -> float | None:
    if len(candles) < 2:
        return None

    trs = []
    for i in range(1, len(candles)):
        high      = candles[i]["high"]
        low       = candles[i]["low"]
        prev_close = candles[i - 1]["close"]
        trs.append(max(high - low, abs(high - prev_close), abs(low - prev_close)))

    slice_ = trs[-period:]
    return sum(slice_) / len(slice_)


def generate_signal(
    spot_price: float,
    price_to_beat: float | None,
    rsi: float | None,
    atr: float | None,
    up_price: float | None,
    down_price: float | None,
) -> dict | None:
    if rsi is None:
        return None

    if atr is not None:
        atr_pct = (atr / spot_price) * 100
        if atr_pct < 0.005 or atr_pct > 0.5:
            return None

    # Mercados "Up or Down": sin umbral de precio, solo RSI + pricing del outcome
    if price_to_beat is None:
        if rsi >= 58 and up_price is not None and up_price < 0.52:
            return {"outcome": "UP"}
        if rsi <= 42 and down_price is not None and down_price < 0.52:
            return {"outcome": "DOWN"}
        return None

    # Mercados con umbral explícito (ej: "Will BTC be above $82,000?")
    divergence = (spot_price - price_to_beat) / price_to_beat

    if rsi >= 58 and divergence > 0.0005 and up_price is not None and up_price < 0.52:
        return {"outcome": "UP"}
    if rsi <= 42 and divergence < -0.0005 and down_price is not None and down_price < 0.52:
        return {"outcome": "DOWN"}

    return None
